blog-content/articles/save-your-future-self-by-encapsulating-test-input-preparation.md

```diff
@@ -0,0 +1,97 @@
+Two numeric inputs and we are done: 
+
+:include-file: com/example/portfolio/SimpleCalculatorTest.java {title: "SimpleCalculatorTest.java"}
+
+Unfortunately real apps require real domain, and real domain is rarely represented by a couple of numbers.
+
+Let's take an example of a fake `ProfitCalculator`. 
+
+Business explained to us how to calculate a potential profit based on a list of `Transactions`. Let's implement
+and test this crucial piece of our software. 
+
+`Transaction` is a simple data class with a few fields (number of fields is reduced for simplicity).
+
+:include-file: com/example/portfolio/Transaction.java {title: "Transaction.java", readMore: true, readMoreVisibleLines: 20}
+
+Here is the first `ProfitCalculator` implementation in all its glory, ready to be tested.
+
+:include-file: com/example/portfolio/ProfitCalculator.java {title: "ProfitCalculator.java"}
+
+And here is the first, boilerplate version of our test.
+
+:include-file: com/example/portfolio/ProfitCalculatorWithoutEncapsulationTest.java {title: "ProfitCalculatorTest.java"}
+
+You may think "test is not that bad", only a few setters to call, not a big deal. 
+The number of setters will only grow as you dig more into requirements. 
+The number of transactions will grow as we deal with more complex scenarios. 
+But that's not all the reasons to be worried.
+
+Another reason to worry is the way `Transaction` instance is being constructed. Right now it is a bunch of setters.
+Tomorrow it becomes a builder pattern, or straight constructor initialization.  
+
+From `ProfitCalculator` perspective the way `Transaction` instances are created is irrelevant as it doesn't affect the calculation logic.
+
+`Transaction` is going to be used in other places and not just `ProfitCalculator`. There will be other tests creating `Transaction` instances.
+The more tests you have, the more reluctant you will be to refactor anything about `Transaction` definition.
+
+# Test Input Initial Encapsulation
+
+Let's extract `Transaction` creation code into `TestTransactions`
+
+:include-file: com/example/portfolio/ProfitCalculatorWithBasicEncapsulationTest.java {title: "ProfitCalculatorTest.java with initial encapsulation"}
+
+:include-java: com/example/portfolio/TestTransactions.java {title: "TestTransactions.createTransaction", entry: "createTransaction"}
+
+Less boilerplate and a good start.
+
+Question: Can you still spot the problem?
+
+--TODO-- add reveal plugin to hide the answer and show it only on a click
+
+As we add more properties to `Transaction` class, some of the tests may need to set extra transaction properties. 
+Additionally some of the properties will be completely irrelevant to the business logic under test, yet you will be forced to set them anyway.
+
+# Test Input Encapsulation With Webtau Table Data   
+
+I was dealing with test problems like above for awhile and eventually came up with a solution that fits my needs.
```

Comment:
I would replace *awhile* with something more concrete, like *for years*, or *for much of my career*